File: apps/api/test/interop.test.ts
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, resolve } from 'node:path';
import { describe, expect, it } from 'vitest';

import {
  fromBase64,
  toBase64,
  sealPayload,
  openPayload,
  derivePublicKeyValue,
  type DeviceKeyMaterial,
} from './reference-client';
import { encryptedPayloadEnvelopeSchema } from '@enclave/contracts';

const VECTOR_PATH = resolve(
  dirname(fileURLToPath(import.meta.url)),
  '../../../packages/contracts/test-vectors/interop-v1.json',
);

const vector = JSON.parse(readFileSync(VECTOR_PATH, 'utf8'));
const parsedEnvelope = encryptedPayloadEnvelopeSchema.parse(vector.envelope);

const material: DeviceKeyMaterial = {
  privateKey: fromBase64(vector.privateKey),
  publicKeyValue: vector.publicKey,
  sealingKey: fromBase64(vector.sealingKey),
  keyVersion: vector.keyVersion,
};

/**
 * Interop assertion strategy (no independent random values):
 * The committed vector's ciphertext was produced by the TS reference client
 * (noble). The Dart suite asserts its seal() output is BYTE-IDENTICAL to that
 * ciphertext with the same fixed nonce, and its open() recovers the plaintext.
 * Here TS re-encrypts with the same fixed nonce and must produce byte-identical
 * output, and TS open() must recover the plaintext. Together:
 *   - TS seal == vector == Dart seal  (both directions byte-equal)
 *   - TS open(vector) == Dart open(vector) == plaintext  (both can decrypt)
 */
describe('cross-language interop (vector: packages/contracts/test-vectors/interop-v1.json)', () => {
  it('TS private key derives the vector X25519 public key', () => {
    expect(derivePublicKeyValue(material.privateKey)).toBe(vector.publicKey);
  });

  it('TS seal with fixed nonce is byte-identical to the vector ciphertext', () => {
    const envelope = sealPayload(
      material,
      vector.deviceId,
      fromBase64(vector.plaintext),
      fromBase64(vector.nonce),
    );
    expect(envelope).toEqual(parsedEnvelope);
  });

  it('TS opens the vector ciphertext (== Dart seal output) to the plaintext', () => {
    const opened = openPayload(material, parsedEnvelope);
    expect(toBase64(opened)).toBe(vector.plaintext);
  });
});

File: packages/dart_core/test/interop_test.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:enclave_crypto/enclave_crypto.dart';
import 'package:test/test.dart';

/// Loads the committed cross-language interop vector.
/// Tests run with cwd = packages/dart_core; the vector lives in
/// packages/contracts/test-vectors/interop-v1.json.
Map<String, dynamic> _loadVector() {
  final path = '../contracts/test-vectors/interop-v1.json';
  if (!File(path).existsSync()) {
    fail('vector not found at $path (run tests from packages/dart_core)');
  }
  return jsonDecode(File(path).readAsStringSync()) as Map<String, dynamic>;
}

void main() {
  final vector = _loadVector();

  DeviceKeyMaterial material() => DeviceKeyMaterial(
        privateKey: Uint8List.fromList(base64Decode(vector['privateKey'] as String)),
        publicKeyValue: vector['publicKey'] as String,
        sealingKey: Uint8List.fromList(base64Decode(vector['sealingKey'] as String)),
        keyVersion: vector['keyVersion'] as int,
      );

  group('cross-language interop (vector: packages/contracts/test-vectors/interop-v1.json)', () {
    test('Dart derives the vector X25519 public key from the private key', () async {
      final derived = await derivePublicKeyValue(
        Uint8List.fromList(base64Decode(vector['privateKey'] as String)),
      );
      expect(derived, vector['publicKey']);
    });

    test('Dart seal with fixed nonce is byte-identical to vector (TS) ciphertext', () async {
      final envelope = await sealPayload(
        material(),
        vector['deviceId'] as String,
        base64Decode(vector['plaintext'] as String),
        Uint8List.fromList(base64Decode(vector['nonce'] as String)),
      );
      expect(envelope.toJson(), vector['envelope']);
    });

    test('Dart opens the vector ciphertext (== TS seal output) to the plaintext', () async {
      final opened = await openPayload(
        material(),
        (vector['envelope'] as Map).cast<String, dynamic>(),
      );
      expect(base64Encode(opened), vector['plaintext']);
    });
  });
}